Index: tools/quake3/q3map2/game_quake3.h
===================================================================
--- tools/quake3/q3map2/game_quake3.h.orig	2008-09-11 10:59:01.000000000 +0200
+++ tools/quake3/q3map2/game_quake3.h	2008-09-11 10:59:07.000000000 +0200
@@ -114,7 +114,7 @@
 	1.0f,				/* lightmap gamma */
 	1.0f,				/* lightmap compensate */
 	"IBSP",				/* bsp file prefix */
-	47,					/* bsp file version */
+	46,					/* bsp file version */
 	qfalse,				/* cod-style lump len/ofs order */
 	LoadIBSPFile,		/* bsp load function */
 	WriteIBSPFile,		/* bsp write function */
